File: scripts/analysis/debugAnalyzer.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Data::Dumper;

# Extract information from debug logs and export to CSV for further analysis.
# Andrew Benson (22-August-2023)

# Get the name of the debug log file.
die("Usage: debugAnalyzer.pl <debugFile>")
    unless ( scalar(@ARGV) == 1 );
my $debugFileName = $ARGV[0];

# Initialize step counter and data.
my $i = -1;
my $data;

# Open and parse the debug log file.
open(my $debugLog,$debugFileName);
while ( my $line = <$debugLog> ) {
    if ( $line =~ m/^\s*step:\s+([0-9\.eE\+\-]+)/ ) {
	++$i;
	$data->{'time'}->[$i] = $1;
    } elsif ( $line =~ m/^\s*value:\s+([a-zA-Z:]+)\s+([\d\.eE\+\-]+)/ ) {
	$data->{'properties'}->{$1}->{'value'}      ->[$i] = $2;
    } elsif ( $line =~ m/^\s*rate:\s+\(([a-zA-Z0-9_]+)\)\s+([a-zA-Z:\d\[\]]+)\s+([\d\.eE\+\-]+)/ ) {
	$data->{'properties'}->{$2}->{'rate' }->{$1}->[$i] = $3;
    } else {
	print $line;
	die("failed to parse line");
    }


    ## AJB HACK
    last
	if ( $i > 10 );

}
close($debugLog);
my $stepCount = $i+1;

# Output values and rates.
my @values = sort(keys(%{$data->{'properties'}}));
my @rates;
foreach my $property ( @values ) {
    push(@rates,map {{property => $property, function => $_}} sort(keys(%{$data->{'properties'}->{$property}->{'rate'}})));
}
open(my $valueLog,">","debugValues.csv");
print $valueLog "time , ".join(" , ",@values)."\n";
for(my $j=0;$j<$stepCount;++$j) {
    print $valueLog $data->{'time'}->[$j]." , ".join(" , ",map {defined($data->{'properties'}->{$_}->{'value'}->[$j]) ? $data->{'properties'}->{$_}->{'value'}->[$j] : "0.000000E+00"} @values)."\n";
}
close($valueLog);
open(my $ratesLog,">","debugRates.csv");
print $ratesLog "time , ".join(" , ",map {$_->{'property'}.":".$_->{'function'}} @rates)."\n";
for(my $j=0;$j<$stepCount;++$j) {
    print $ratesLog $data->{'time'}->[$j]." , ".join(" , ",map {my $rate = $data->{'properties'}->{$_->{'property'}}->{'rate'}->{$_->{'function'}}->[$j]; defined($rate) ? $rate : "0.000000E+00"} @rates)."\n";
}
close($ratesLog);
exit;

File: scripts/analysis/debugAnalyzer.py
#!/usr/bin/env python3
# Extract information from debug logs and export to CSV for further analysis.
# Andrew Benson (22-August-2023)

import re
import sys
from collections import defaultdict

# Get the name of the debug log file.
if len(sys.argv) != 2:
    sys.exit("Usage: debugAnalyzer.py <debugFile>")
debug_file_name = sys.argv[1]

# Initialize step counter and data.
i = -1
times = []
# properties[name]['value'][step] = value
# properties[name]['rate'][function][step] = rate
properties = defaultdict(lambda: {'value': {}, 'rate': defaultdict(dict)})

# Open and parse the debug log file.
re_step  = re.compile(r'^\s*step:\s+([0-9\.eE\+\-]+)')
re_value = re.compile(r'^\s*value:\s+([a-zA-Z:]+)\s+([\d\.eE\+\-]+)')
re_rate  = re.compile(r'^\s*rate:\s+\(([a-zA-Z0-9_]+)\)\s+([a-zA-Z:\d\[\]]+)\s+([\d\.eE\+\-]+)')

try:
    with open(debug_file_name) as debug_log:
        for line in debug_log:
            m = re_step.match(line)
            if m:
                i += 1
                times.append(m.group(1))
            else:
                m = re_value.match(line)
                if m:
                    properties[m.group(1)]['value'][i] = m.group(2)
                else:
                    m = re_rate.match(line)
                    if m:
                        properties[m.group(2)]['rate'][m.group(1)][i] = m.group(3)
                    else:
                        print(line, end='')
                        sys.exit("failed to parse line")
except OSError as e:
    sys.exit(f"failed to open {debug_file_name}: {e.strerror or e}")

step_count = i + 1

# Output values and rates.
prop_names = sorted(properties.keys())
rates = [
    {'property': prop, 'function': fn}
    for prop in prop_names
    for fn in sorted(properties[prop]['rate'].keys())
]

MISSING = "0.000000E+00"

with open("debugValues.csv", "w") as value_log:
    value_log.write("time , " + " , ".join(prop_names) + "\n")
    for j in range(step_count):
        values = [properties[p]['value'].get(j, MISSING) for p in prop_names]
        value_log.write(times[j] + " , " + " , ".join(values) + "\n")

with open("debugRates.csv", "w") as rates_log:
    headers = [r['property'] + ":" + r['function'] for r in rates]
    rates_log.write("time , " + " , ".join(headers) + "\n")
    for j in range(step_count):
        row = [properties[r['property']]['rate'][r['function']].get(j, MISSING) for r in rates]
        rates_log.write(times[j] + " , " + " , ".join(row) + "\n")
